### PR TITLE
Clearer Health and Vitamin Fixes

### PR DESCRIPTION
**Pull request by bombasticSlacks**
#### Summary
Interface "Clearer Health in UI"

#### Purpose of change
Should be able to track daily vitamins since it's now important for health. Should be able to check in on how your character is feeling always not just in the mornings.

While working on this realize daily vitamins don't work (for some vitamins).

#### Describe the solution
This PR Does 3 things:

1. Adds vitamin info to the consume menu so you know how you are doing daily.
Your character can get a guess if they don't have a tracking tool, if they do they get a % (similar to calories with a phone / fitness band). Adds a new daily counter for vitamins that is a guess (not your actual digested amount). So while you are working on that orange in your stomach you still counted its vit c an hour ago.

2. Adds Health info (as a guess to the status menu)

3. Fixes broken vitamins
Some of the vitamins had issues where they could never be positive so they never gave health benefits. I've added a new tracker for daily intake (similar to all the other daily intakes). This is manipulated when digestion happens and is checked every 24h for the health benefits. 

4. Basically all vitamins were broken in at least 1 situation for each way they could be used. This is because Food had vitamins passed in in RDA but it was never converted to the games internal units. So now it properly is. Medicines also no longer do a weird double convert. I've added some really simple clean unit tests to make sure that if you eat something with 100% RDA you are in equilibrium for a day. 
5. Moved vitamins back to around 100 units per day adjusting everything else to basically match. This is where the original rate of 15 minutes per unit comes from (96 per day). I've still got my math to get it closer to "perfect" converting to and from RDA BUT now some of the values aren't as massively compressed. Like calcium for example was producing outrageously small numbers for RDA which leads to rounding issues.

#### Testing
I added debug info and this brings a bunch of stuff forward in the UI for testing.

#### Additional context
Stats Menu
image

With Tracker
image

Without Tracker
image

DEBUG INFO 
image

### CODE REVIEW

File: src/player_display.cpp [hunk 1]
```diff
@@ -532,18 +536,22 @@ static void draw_stats_info( const catacurses::window &w_info, const Character &
         fold_and_print( w_info, point( 1, 1 + lines ), FULL_SCREEN_WIDTH - 2, c_light_gray,
                         display::weight_long_description( you ) );
     } else if( line == 5 ) {
+        // NOLINTNEXTLINE(cata-use-named-point-constants)
+        const int lines = fold_and_print( w_info, point( 1, 0 ), FULL_SCREEN_WIDTH - 2, c_magenta,
+                                          _( "How healthy you feel.  Excercise, vitamins, sleep and not ingesting poison will increase this overtime." ) );
```
Comment (pjf) [inline src/player_display.cpp:541]: Are we missing a `fold_and_print` or similar here? I can't see `lines` being used anywhere. (Neither could my compiler)

File: src/player_display.cpp [hunk 2]
```diff
@@ -532,18 +536,22 @@ static void draw_stats_info( const catacurses::window &w_info, const Character &
         fold_and_print( w_info, point( 1, 1 + lines ), FULL_SCREEN_WIDTH - 2, c_light_gray,
                         display::weight_long_description( you ) );
     } else if( line == 5 ) {
+        // NOLINTNEXTLINE(cata-use-named-point-constants)
+        const int lines = fold_and_print( w_info, point( 1, 0 ), FULL_SCREEN_WIDTH - 2, c_magenta,
```
Comment (pjf) [inline src/player_display.cpp:540]: My compiler, too, is unhappy we're not using `lines` anywhere.

Calling `fold_and_print` without capturing the return value seems to be compiling fine, though. :)

File: src/vitamin.h [hunk 3]
```diff
@@ -105,6 +105,8 @@ class vitamin
         efftype_id excess_;
         int min_ = 0;
         int max_ = 0;
+        // how much you've taken today
+        int daily_count = 0;
```
Comment (pjf) [inline src/vitamin.h:109]: Uh oh, I can't spot where `daily_count` gets used!

**Comment by pjf** [comment]
Running this locally to playtest (after making a few trivial changes in my review to appease the compiler), and so far it's working great, *except* the tracker shows twice the amount of iron I've consumed (according to the food descriptions), and calcium shows *six times* that.

Eg, a slice of `humble pie` made with sweetbreads lists about a 20% daily intake of iron, but I end up with 40% being shown on the tracker.

(In this case the tracker is the smartphone nutrition app)

**Comment by pjf** [comment]
I'm going to guess the weird 2x iron and 6x calcium has to do with iron and calcium having 2x and 6x the `rate` of Vitamin C in `vitamin.json`.

**Review by pjf — CHANGES_REQUESTED** [review]
- Compiler is picky about a few lines
- Tracking app has 2x and 6x multiplier for iron and calcium, likely due to `rate` in `vitamin.json`.

**Comment by bombasticSlacks** [comment]
> I'm going to guess the weird [2x](https://github.com/bombasticSlacks/Cataclysm-DDA/blob/f609307fe480f91a8644eb77eb7c6122496fd017/data/json/vitamin.json#L20) iron and [6x](https://github.com/bombasticSlacks/Cataclysm-DDA/blob/f609307fe480f91a8644eb77eb7c6122496fd017/data/json/vitamin.json#L9) calcium has to do with iron and calcium having 2x and 6x the rate of Vitamin C in vitamin.json.

I duplicated the RDA code used to calculate health values so this is exactly the kind of stuff I was hoping to spot by making this visible to players. 

I'll standardized where RDA is calculated so that it shows identically across examine and what it actually gives.

**Comment by pjf** [comment]
> I duplicated the RDA code used to calculate health values so this is exactly the kind of stuff I was hoping to spot by making this visible to players.

I am delighted to be spotting this!

Thank you *so much* for this PR, it's a huge quality of life improvement!

File: src/player_display.cpp (hunk 2 above)
Comment (pjf) [inline src/player_display.cpp:540]:
```suggestion
        fold_and_print( w_info, point( 1, 0 ), FULL_SCREEN_WIDTH - 2, c_magenta,
```

Comment (pjf) [reply src/player_display.cpp:540]:
```suggestion
        fold_and_print( w_info, point( 1, 0 ), FULL_SCREEN_WIDTH - 2, c_magenta,
```

File: src/consumption.cpp [hunk 4]
```diff
@@ -589,6 +592,27 @@ int Character::vitamin_get( const vitamin_id &vit ) const
...
+int Character::get_daily_vitamin( const vitamin_id &vit, bool actual ) const
+{
+    if( get_option<bool>( "NO_VITAMINS" ) && vit->type() == vitamin_type::VITAMIN ) {
+        return 0;
+    }
+
+    const auto &v = daily_vitamins.find( vit );
+    // if we should get the guess or the real value
+    const int &count = actual ? v->second.second : v->second.first;
+    return v != daily_vitamins.end() ? count : 0;
```
Comment (BevapDin) [inline src/consumption.cpp:604]: If v is an iterator to the end, dereferencing it triggers undefined behaviour. And you dereference it in the previous line before that check.

File: src/debug_menu.cpp [hunk 5]
```diff
@@ -1310,7 +1310,8 @@ static void character_edit_needs_menu( Character &you )
 
     const auto &vits = vitamin::all();
     for( const auto &v : vits ) {
-        smenu.addentry( -1, true, 0, "%s: %d", v.second.name(), you.vitamin_get( v.first ) );
+        smenu.addentry( -1, true, 0, "%s: daily %d, overall %d", v.second.name(),
```
Comment (BevapDin) [inline src/debug_menu.cpp:1313]: Needs translation:
```suggestion
        smenu.addentry( -1, true, 0, _( "%s: daily %d, overall %d" ), v.second.name(),
```

File: src/character.h [hunk 6]
```diff
@@ -3439,6 +3444,10 @@ class Character : public Creature, public visitable
         move_mode_id move_mode;
         /** Current deficiency/excess quantity for each vitamin */
         std::map<vitamin_id, int> vitamin_levels;
+        /** Current quantity for each vitamin today first value is expected already in % (for UI) second value is actual (digested) in vitamin units*/
```
Comment (bombasticSlacks) [inline src/character.h:3447]:
```suggestion
        /** Current quantity for each vitamin today first value is expected second value is actual (digested) in vitamin units*/
```

File: tests/stomach_contents_test.cpp [hunk 7]
```diff
@@ -132,6 +141,135 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+    // check
+    CHECK( subject.vitamin_get( vitamin_iron ) <= -95 );
+    CHECK( subject.vitamin_get( vitamin_calcium ) <= -95 );
+    CHECK( subject.vitamin_get( vitamin_vitC ) <= -95 );
+    CHECK( subject.vitamin_get( vitamin_iron ) >= -97 );
+    CHECK( subject.vitamin_get( vitamin_calcium ) >= -97 );
+    CHECK( subject.vitamin_get( vitamin_vitC ) >= -97 );
+
+    // slow vitamin drains every 90 minutes or 16 units in a day
+    CHECK( subject.vitamin_get( vitamin_vitC ) <= -15 );
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:171]:
```suggestion
    CHECK( subject.vitamin_get( vitamin_slow ) <= -15 );
```

File: tests/stomach_contents_test.cpp [hunk 8]
```diff
@@ -132,6 +141,135 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+    CHECK( subject.vitamin_get( vitamin_vitC ) >= -17 );
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:172]:
```suggestion
    CHECK( subject.vitamin_get( vitamin_slow ) >= -17 );
```

File: tests/stomach_contents_test.cpp [hunk 9]
```diff
@@ -132,6 +141,135 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+
+    // fast vitamin drains every 5 minutes or 288 units in a day
+    CHECK( subject.vitamin_get( vitamin_vitC ) <= -287 );
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:175]:
```suggestion
    CHECK( subject.vitamin_get( vitamin_fast ) <= -287 );
```

File: tests/stomach_contents_test.cpp [hunk 10]
```diff
@@ -132,6 +141,135 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+    CHECK( subject.vitamin_get( vitamin_vitC ) >= -289 );
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:176]:
```suggestion
    CHECK( subject.vitamin_get( vitamin_fast ) >= -289 );
```

File: tests/stomach_contents_test.cpp [hunk 11]
```diff
@@ -132,6 +141,143 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+    REQUIRE( subject.get_daily_vitamin( vitamin_calcium ) != 0 );
+    REQUIRE( subject.get_daily_vitamin( vitamin_iron ) != 0 );
+    pass_time( subject, 12_hours );
+
+    // check if after a day health is up and vitamins are reset
+    CHECK( subject.get_daily_vitamin( vitamin_vitC ) == 0 );
+    CHECK( subject.get_daily_vitamin( vitamin_calcium ) == 0 );
+    CHECK( subject.get_daily_vitamin( vitamin_iron ) == 0 );
+    // get that vitamin health bonus +/- whatever other maluses applied in a day
+    CHECK( subject.get_daily_health() >= 6 );
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:277]:
```suggestion
    CHECK( subject.get_daily_health() >= 5 );
```

File: tests/stomach_contents_test.cpp [hunk 12]
```diff
@@ -132,6 +141,149 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+        subject.set_daily_health( 0 );
+
+    }
+
+    // check if after a day health is up and vitamins are reset
+    CHECK( subject.get_daily_vitamin( vitamin_vitC ) == 0 );
+    CHECK( subject.get_daily_vitamin( vitamin_calcium ) == 0 );
+    CHECK( subject.get_daily_vitamin( vitamin_iron ) == 0 );
+    // get that vitamin health bonus is up by 6
+    CHECK( subject.get_daily_health() >= 6 );
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:283]:
```suggestion
    CHECK( subject.get_daily_health() >= 5 );
```

File: tests/stomach_contents_test.cpp [hunk 13]
```diff
@@ -132,6 +141,149 @@ TEST_CASE( "starve_test", "[starve][slow]" )
...
+        //otherwise clean up any other health changes that may have happened
+        subject.set_daily_health( 0 );
+
+    }
+
+    // check if after a day health is up and vitamins are reset
+    CHECK( subject.get_daily_vitamin( vitamin_vitC ) == 0 );
+    CHECK( subject.get_daily_vitamin( vitamin_calcium ) == 0 );
+    CHECK( subject.get_daily_vitamin( vitamin_iron ) == 0 );
+    // get that vitamin health bonus is up by 6
```
Comment (bombasticSlacks) [inline tests/stomach_contents_test.cpp:282]:
```suggestion
    // get that vitamin health bonus is up by 6 with maybe a recent reduction on the same timecheck
```

File: src/consumption.cpp [hunk 14]
```diff
@@ -561,6 +569,9 @@ int Character::vitamin_mod( const vitamin_id &vit, int qty )
         it->second = std::min( it->second + qty, v.max() );
         update_vitamins( vit );
 
+        // update the daily trackers too while here
+        daily_vitamins[vit].second += qty;
```
Comment (akrieger) [inline src/consumption.cpp:573]: mac ubsan tests are failing here
```
(~[slow] ~[.],starting_items)=> src/consumption.cpp:573:36: runtime error: signed integer overflow: 2147483647 + 2147481247 cannot be represented in type 'int'
```

Comment (akrieger) [reply src/consumption.cpp:573]: Tests caught it too.
https://github.com/CleverRaven/Cataclysm-DDA/runs/7663399494?check_suite_focus=true#step:16:329

Comment (dseguin) [reply src/consumption.cpp:573]: To be fixed in #59952, looks like

**Comment by bombasticSlacks** [comment]
Sorry y'all wrestled with tests so long on this one that I missed it.

**Pull request merged**